Added macvlan and ipvlan drivers

nerdalert opened the pull request:
- Notes at: https://gist.github.com/nerdalert/c0363c15d20986633fda

Signed-off-by: Brent Salisbury <brent@docker.com>

aboch commented on drivers/ipvlan/ipvlan_test.go:
```diff
@@ -0,0 +1,59 @@
+import (
+	"testing"
+
+	"github.com/docker/libnetwork/driverapi"
```
You need to add this import for the CI to pass:
`_ "github.com/docker/libnetwork/testutils"`

nerdalert replied:
Thanks fellas! @aboch @mavenugo that fixed it.

mavenugo commented:
@nerdalert :clap: thanks for the contribution. We will review it and provide feedback.

aboch commented on drivers/ipvlan/ipvlan_state.go:
```diff
@@ -0,0 +1,110 @@
+	d.Lock()
+	networks := d.networks
+	d.Unlock()
+	n, ok := networks[nid]
```
We don't need `networks`, we only need to lock around this line only, 

```
d.Lock()
n, ok := d.networks[nid]
d.Unlock()
```

aboch commented on drivers/ipvlan/ipvlan_network.go:
```diff
@@ -0,0 +1,163 @@
+	for label, value := range labels {
+		switch label {
+		case hostIfaceOpt:
+			// parse driver option '-o --ipvlan_mode'
```
I think this comment and the one below should be swapped.

aboch commented on drivers/ipvlan/ipvlan_joinleave.go:
```diff
@@ -0,0 +1,105 @@
+	endpoint.srcName = vethName
+	if n.netConfig.IpvlanMode == modeL3 {
+		// disable gateway services to add a default gw using dev eth0 only
+		jinfo.DisableGatewayService()
```
This is not needed if you are anyway adding a default static route to joininfo.

nerdalert replied:
Great point, what I ran into is If I don't disable the service, gateway services will try and add a default route on top of the static default route that is added. The route is added as a static because L3 mode doesn't have a next hop and looks like this to the host `default dev eth0`.

`failed to set gateway while updating gateway: file exists` complete logs [here](https://gist.github.com/884e320817d31d48a636).

aboch replied:
You should no longer run into that error, see: https://github.com/docker/libnetwork/blob/master/default_gateway.go#L117

aboch replied:
Unless that check does not work, which seems to be the case... I will take a look at it.

aboch replied:
Ok, understand now, endpoint only stores the `types.NEXT_HOP` static routes. This one is of type `types.CONNECTED`. Thanks.

aboch commented on drivers/ipvlan/ipvlan_joinleave.go:
```diff
@@ -0,0 +1,105 @@
+		if err != nil {
+			return err
+		}
+		if err := jinfo.AddStaticRoute(defaultRoute.Destination, defaultRoute.RouteType, defaultRoute.NextHop); err != nil {
```
Shouldn't we fail the join in case of this failure ?

nerdalert replied:
Added, ty!

```
if err := jinfo.AddStaticRoute(defaultRoute.Destination, defaultRoute.RouteType, defaultRoute.NextHop); err != nil {
	return fmt.Errorf("failed to set an ipvlan l3 mode default gateway: %v", err)
}
```

aboch commented on drivers/ipvlan/ipvlan.go:
```diff
@@ -0,0 +1,85 @@
+	sbox      osl.Sandbox
+	endpoints endpointTable
+	driver    *driver
+	netConfig *networkConfiguration
```
Can we simply call it `config` being it  already a member of `network` ?

aboch commented on drivers/ipvlan/ipvlan_network.go:
```diff
@@ -0,0 +1,163 @@
+	}
+	// verify the macvlan mode from --ipvlan_mode option
+	switch config.IpvlanMode {
+	case modeL2:
```
If you think it is worth to make the cange, FYI golang would allow:
```
case modelL2, "":
    // default to ipvlan L2 mode if --ipvlan_mode is empty
    config.IpvlanMode = modeL2
```

aboch commented on drivers/macvlan/macvlan_network.go:
```diff
@@ -0,0 +1,170 @@
+	for label, value := range labels {
+		switch label {
+		case hostIfaceOpt:
+			// parse driver option '-o --macvlan_mode'
```
comments are swapped

aboch commented on drivers/ipvlan/ipvlan_setup.go:
```diff
@@ -0,0 +1,177 @@
+
+// validateHostIface check if the specified interface exists in the default namespace
+func validateHostIface(ifaceStr string) bool {
+	_, err := net.InterfaceByName(ifaceStr)
```
Can you replace this with `netlink.LinkbyName()`
There is a considerable performance improvement based on https://github.com/LK4D4/libnetwork/commit/a815036c2fd730b909508c1fc0a89db030bafbf6

nerdalert replied:
Excellent, I use `net` for iface checks thinking it would be more performant. Great to know, ty.

aboch commented on drivers/macvlan/macvlan_setup.go:
```diff
@@ -0,0 +1,187 @@
+
+// validateHostIface check if the specified interface exists in the default namespace
+func validateHostIface(ifaceStr string) bool {
+	_, err := net.InterfaceByName(ifaceStr)
```
Same here, can you replace this with netlink.LinkbyName()

aboch commented on drivers/ipvlan/ipvlan.go:
```diff
@@ -0,0 +1,85 @@
+	ipvlanType          = "ipvlan"     // driver type name
+	modeL2              = "l2"         // ipvlan mode l2 is the default
+	modeL3              = "l3"         // ipvlan L3 mode
+	hostIfaceOpt        = "host_iface" // host interface -o UX key
```
I am not sure if it still stands, other can comment: Shouldn't the driver specific options still follow the `com.docker.network.<driver name>.<option name>` format ?

aboch commented on drivers/macvlan/macvlan_errors.go:
```diff
@@ -0,0 +1,227 @@
+package macvlan
```
We do not really need typed errors. Actually trying to move away from them. If you check overlay driver, it is not using them. You can just use the `types.<errors>` in case you need to return something the caller needs to evaluate, for example bridge driver to returns a MaskableError in some instances. One day we'll remove the typed errors from bridge driver as well.
(Also I see not all of the following are used in the code)

aboch commented on drivers/ipvlan/ipvlan_errors.go:
```diff
@@ -0,0 +1,227 @@
+package ipvlan
```
Same here, I don't think we need this file.

aboch commented on drivers/ipvlan/ipvlan_network.go:
```diff
@@ -0,0 +1,163 @@
+			return nil, err
+		}
+	}
+	return config, nil
```
Can the `netlabel.Internal` option be honored with this driver ?
If not, we should fail the network creation if that option was passed.

aboch commented on drivers/ipvlan/ipvlan_network.go:
```diff
@@ -0,0 +1,174 @@
+		err = createVlanSubIface(config.HostIface)
+		if err != nil {
+			return err
+		}
```
With current drivers inter-network L3 traffic is not allowed. IPVlan networks should conform to this behavior.
But I see that containers on two L3 mode ipvlan networks sharing the same parent interface, can ping each other.

I think we should block the second network creation, if there is no other way to keep the two networks separated.

nerdalert replied:
Thanks for the thoughts, I totally get your point on isolating them. On the other hand, the ability for disparate ip prefixes to be reachable without an explicit router/gateway is what makes L3 mode interesting by breaking the mold of viewing a network as an ip subnet. 

- Putting up two examples for clarity for folks.

**Example #2:** Creates networks on the same shared parent interface `-o host_iface=eth0.99` where the two hosts can talk to one another. 

- Shared Network Domain - networks sharing the same parent interface `-o host_iface=`

```
### Create one more l3 network using the same parent iface as the previous network with a different subnet
docker network create -d ipvlan --subnet=192.168.91.0/24 -o host_iface=eth0.99 -o ipvlan_mode=l3 backend_net1
docker network create -d ipvlan --subnet=192.168.99.0/24 -o host_iface=eth0.99 -o ipvlan_mode=l3 backend_net2

### Start containers on backend_net1
docker run --net=backend_net1 --name=backend_host --ip=192.168.91.10 -itd alpine /bin/sh

# Do a test ping from the next host to 192.168.99.10 via 192.168.91.10
# *Note* this works across L3 boundaries because you share the 
# same parent interface (eth0.99) even though it is across discontiguous subnets
docker run --net=backend_net2 --ip=192.168.99.10 -it --rm alpine /bin/sh

# Containers on the SAME parent interface (-o host_iface=) CAN ping one another even across L3 boundaries (e.g. subnets)
/ # ping  192.168.91.10
PING 192.168.91.10 (192.168.91.10): 56 data bytes
64 bytes from 192.168.91.10: seq=0 ttl=64 time=0.060 ms
64 bytes from 192.168.91.10: seq=1 ttl=64 time=0.078 ms
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.060/0.069/0.078 ms

```

**Example #2:** creates networks on two different parent interfaces  `-o host_iface=eth0.94` & `-o host_iface=eth0.95` which prevent the slave interfaces in the containers from reaching one another, unless there was an external routing process to hairpin the traffic between containers which is beyond the scope of the driver.


- Isolated Network: networks not sharing the same parent  interface `-o host_iface=`

```
### Same network prefixes but with 
docker network create -d ipvlan --subnet=192.168.94.0/24 -o host_iface=eth0.94 -o ipvlan_mode=l3 web_net
docker network create -d ipvlan --subnet=192.168.95.0/24 -o host_iface=eth0.95 -o ipvlan_mode=l3 db_net

### Start containers on each network
docker run --net=web_net --ip=192.168.94.10 --name=web_test2 -itd alpine /bin/sh


# Containers on DIFFERENT parent interface (-o host_iface=) can NOT ping one another even across L3 boundaries (e.g. subnets)
docker run --net=db_net -it --rm alpine /bin/sh
ping 192.168.95.10
PING 192.168.95.10 (192.168.95.10): 56 data bytes
--- 192.168.95.10 ping statistics ---
2 packets transmitted, 0 packets received, 100% packet loss
```

To add a bit of context for others not familiar with ipvlan l3 mode is assigning each container host routes `/32` for mobility from subnet/l2 locality. L3 needs a route to be distributed throughout a cluster to be functional whether its integrated with the network fabric IGP or in an overlay with a distributed datastore etc. I'm thinking implicitly isolating based on IP subnet may limit some of the intended user/3rd party plugin use cases.

chenchun replied:
> I'm thinking implicitly isolating based on IP subnet may limit some of the intended user/3rd party plugin use cases.

This is not true. I think @aboch does not suggest to provide inter-network isolation based on IP subnet. Besides, a single network may consist of multiple subnets. What he suggest is that we should fail creating more networks if we can't provide isolation between them. 

If the IPVlan driver is designed to support vlan, I think a vlan id per network is the right way to provide isolation.

nerdalert replied:
multi-subnet sounds reasonable to me. Regarding `If the IPVlan driver is designed to support vlan` It's worth keeping host routes in mind since untagged traffic is also part of the equation in purely routed environment.

aboch commented on drivers/ipvlan/ipvlan_endpoint.go:
```diff
@@ -0,0 +1,60 @@
+			return err
+		}
+	}
+	n.addEndpoint(ep)
```
I do not see any code for handling the port mapping, in case user starts the container specifying the exposed ports.
If this is not available yet, we should refuse the endpoint creation with proper error message.

nerdalert replied:
Thanks for the thoughts, I totally get your point on isolating them. On the other hand, the ability for disparate ip prefixes to be reachable without an explicit router/gateway is what makes L3 mode interesting by breaking the mold of viewing a network as an ip subnet. 

- Putting up two examples for clarity for folks.

**Example #2:** Creates networks on the same shared parent interface `-o host_iface=eth0.99` where the two hosts can talk to one another. 

- Shared Network Domain - networks sharing the same parent interface `-o host_iface=`

```
### Create one more l3 network using the same parent iface as the previous network with a different subnet
docker network create -d ipvlan --subnet=192.168.91.0/24 -o host_iface=eth0.99 -o ipvlan_mode=l3 backend_net1
docker network create -d ipvlan --subnet=192.168.99.0/24 -o host_iface=eth0.99 -o ipvlan_mode=l3 backend_net2

### Start containers on backend_net1
docker run --net=backend_net1 --name=backend_host --ip=192.168.91.10 -itd alpine /bin/sh

# Do a test ping from the next host to 192.168.99.10 via 192.168.91.10
# *Note* this works across L3 boundaries because you share the 
# same parent interface (eth0.99) even though it is across discontiguous subnets
docker run --net=backend_net2 --ip=192.168.99.10 -it --rm alpine /bin/sh

# Containers on the SAME parent interface (-o host_iface=) CAN ping one another even across L3 boundaries (e.g. subnets)
/ # ping  192.168.91.10
PING 192.168.91.10 (192.168.91.10): 56 data bytes
64 bytes from 192.168.91.10: seq=0 ttl=64 time=0.060 ms
64 bytes from 192.168.91.10: seq=1 ttl=64 time=0.078 ms
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 0.060/0.069/0.078 ms

```

**Example #2:** creates networks on two different parent interfaces  `-o host_iface=eth0.94` & `-o host_iface=eth0.95` which prevent the slave interfaces in the containers from reaching one another, unless there was an external routing process to hairpin the traffic between containers which is beyond the scope of the driver.


- Isolated Network: networks not sharing the same parent  interface `-o host_iface=`

```
### Same network prefixes but with 
docker network create -d ipvlan --subnet=192.168.94.0/24 -o host_iface=eth0.94 -o ipvlan_mode=l3 web_net
docker network create -d ipvlan --subnet=192.168.95.0/24 -o host_iface=eth0.95 -o ipvlan_mode=l3 db_net

### Start containers on each network
docker run --net=web_net --ip=192.168.94.10 --name=web_test2 -itd alpine /bin/sh


# Containers on DIFFERENT parent interface (-o host_iface=) can NOT ping one another even across L3 boundaries (e.g. subnets)
docker run --net=db_net -it --rm alpine /bin/sh
ping 192.168.95.10
PING 192.168.95.10 (192.168.95.10): 56 data bytes
--- 192.168.95.10 ping statistics ---
2 packets transmitted, 0 packets received, 100% packet loss
```

To add a bit of context for others not familiar with ipvlan l3 mode is assigning each container host routes `/32` for mobility from subnet/l2 locality. L3 needs a route to be distributed throughout a cluster to be functional whether its integrated with the network fabric IGP or in an overlay with a distributed datastore etc. I'm thinking implicitly isolating based on IP subnet may limit some of the intended user/3rd party plugin use cases.

chenchun commented on drivers/ipvlan/ipvlan_network.go:
```diff
@@ -0,0 +1,174 @@
+		return err
+	}
+	// update persistent db with the new network
+	return d.storeUpdate(config)
```
I know bridge driver does it the same way, but if `d.storeUpdate(config)` returns error,  we don't rollback the changes of `d.createNetwork(config)`.

nerdalert replied:
Great point. Rolled it back deleting `*network` on err.

```
	err = d.createNetwork(config)
	if err != nil {
		return err
	}
	// update persistent db, rollback on fail
	err = d.storeUpdate(config)
	if err != nil {
		d.deleteNetwork(config.ID)
		return err
	}
	return nil
```

chenchun commented on drivers/ipvlan/ipvlan_network.go:
```diff
@@ -0,0 +1,174 @@
+	if err != nil {
+		return err
+	}
+	if id == "" {
```
Do we have to explicitly check if id is empty ?

chenchun commented on drivers/ipvlan/ipvlan_network.go:
```diff
@@ -0,0 +1,174 @@
+
+// DeleteNetwork the network for the specified driver type
+func (d *driver) DeleteNetwork(nid string) error {
+	if nid == "" {
```
Same here

chenchun commented on drivers/macvlan/macvlan_setup.go:
```diff
@@ -0,0 +1,214 @@
+}
+
+// validateHostIface check if the specified interface exists in the default namespace
+func validateHostIface(ifaceStr string) bool {
```
Lots of methods such as validateHostIface/validateSubIface/createVlanSubIface/addVlanIface/verifyVlanDel are duplicated in ipvlan_setup.go

chenchun commented on drivers/ipvlan/ipvlan_setup.go:
```diff
@@ -0,0 +1,204 @@
+			return fmt.Errorf("unable to parse a valid parent iface: %s and vlan id from: %s (Ex. eth0.10 for vlan 10)", parentIface, vidStr)
+		}
+		// VLAN identifier or VID is a 12-bit field specifying the VLAN to which the frame belongs
+		if vid > 4094 {
```
Can we consolidate these two checks ? if vid > 4094 && vid < 1

chenchun commented on drivers/ipvlan/ipvlan_network.go:
```diff
@@ -0,0 +1,174 @@
+		return fmt.Errorf("could not find network with ID %s", nid)
+	}
+	// if the interface exists, only delete if it matches iface.vlan naming
+	if ok := validateHostIface(n.config.HostIface); ok {
```
If it matches iface.vlan naming, it still can be a user created vlan device and may have other usage. Can we define a variable to conform this?

nerdalert replied:
Im game, to clarify do you mean another `-o vlan_id=`. I initially did but there was some feedback on too many options. I tend to like being explicit since inferring can have conflicts. Currently if it exists it uses it to allow for custom naming but another option keeps things clean.

chenchun replied:
Sorry, I mean if it matches iface.vlan format but not created by libnetwork, we'd better not delete the device here.

chenchun commented on drivers/ipvlan/ipvlan_setup.go:
```diff
@@ -0,0 +1,204 @@
+
+// createVlanSubIface parses sub-interfaces and vlan id for creation
+func createVlanSubIface(ifaceName string) error {
+	if strings.Contains(ifaceName, ".") {
```
Duplicated checks, we have already doing this in `validateSubIface`. I think we can consolidate these two functions. Maybe rename it to ensureSubIface?

mavenugo commented on drivers/macvlan/macvlan_setup.go:
```diff
@@ -0,0 +1,175 @@
+			return nil
+		}
+	}
+	return fmt.Errorf("required kernel module '%s' was not found in /proc/modules, kernel version >= 3.19 is recommended", macvlanType)
```
Is this true ?
I thought macvlan is supported in earlier kernels ?

mavenugo commented:
@nerdalert Given the importance of getting feedback from users, shall we get these drivers under experimental for 1.11 and get enough feedback & also fix the L3 routing use-case. WDYT ?

aboch commented:
Agree with @mavenugo. Being two new drivers, it will take some time with incremental changes to get them where we want them to be.
Introducing them in experimental seems the right approach to me as well.

mavenugo commented:
@nerdalert I understand the need for `-o host_iface`  option in order for the macvlan/ipvlan to plumb to underlay. 
```
$ sudo docker network create -d macvlan test
ERRO[0008] Handler for POST /v1.22/networks/create returned error: macvlan requires an interface from the docker host to be specified (usage: -o host_iface=eth)
Error response from daemon: macvlan requires an interface from the docker host to be specified (usage: -o host_iface=eth)
```

But, as discussed,in the above case, if the user has not specified the host interface, we should consider it as an `--internal` case and automatically create a dummy interface and let the network creation successful. WDYT ?

mavenugo commented:
@nerdalert few more comments as discussed. (commenting it here for better tracking)

* Supporting `--internal` by using `dummy` interface. We can infact implement `--internal` and a configuration without `-o host_iface` similarly with the `dummy` interface approach.
* Handle `--mac-address` option appropriately. It should error out in `ipvlan` but `macvlan` driver must accept and honor it.
* Port-Mapping is infact a tricky scenario for these drivers. I would suggest a warning message and ignoring the option is better than Error'ing out (considering the portability concerns.
* When multiple networks try to share the same parent interface, we should error it out with just the Netlink's `device is busy` message. But if the message doesnt contain the parent interface info, then it is better to customize it by indicating that the user is trying to reuse the parent interface.

nerdalert commented:
@mavenugo great suggestions.
- `--internal` is now supported. If either the `--internal` flag is set or the  `--parent=` interface is not  included then a netlink dummy interface is setup to act as a master parent to attach a macvlan/ipvlan space device to. Once containers are brought up on that network they have full communications with one another but no where outside of the dummy parent interface. 
- Changed port-mapping back to a warning for compose backward compatibility as you mentioned.
- Changed `-o host_iface=` to `-o parent=`.
- One outstanding is setting the MAC addr on Macvlan mode. I will open an issue to track it. Netlink can set a user specified MAC since it is intended to be unique for each interface while Ipvlan uses the same MAC as the parent by design and tracks IP --> port mappings instead of MAC --> port mappings in it's lookup tables.
- Added 50+ networking scenarios for vetting bugs in this script [ipvlan-macvlan-it.sh](https://github.com/nerdalert/dotfiles/blob/master/ipvlan-macvlan-it.sh) that has example usage and capabilities.

mavenugo commented on drivers/ipvlan/ipvlan_network.go:
```diff
@@ -0,0 +1,223 @@
+	}
+	if !parentExists(config.Parent) {
+		// if the parent iface was not specified, create a dummy link
+		if config.Parent == stringid.TruncateID(config.ID) {
```
I think it is better to check for `config.Internal` here.

nerdalert replied:
Done!

mavenugo commented on drivers/ipvlan/ipvlan_network.go:
```diff
@@ -0,0 +1,223 @@
+			}
+			config.CreatedSlaveLink = true
+			// notify the user the did not pass a -o parent= so they know they are --internal
+			if !config.Internal {
```
If you accept to the above comment, then this need to be updated as well.

nerdalert replied:
Done!

mavenugo commented on drivers/macvlan/macvlan_setup.go:
```diff
@@ -0,0 +1,226 @@
+	// create a parent interface since one was not specified
+	parent := &netlink.Dummy{
+		LinkAttrs: netlink.LinkAttrs{
+			Name: dummyName,
```
can you add "macvlan-" as a prefix to the `dummyName` ? it makes it easier for troubleshooting.

nerdalert replied:
I tried that originally but hit the max length of the name. Here are some examples, see what you think:

```
ip link add macvlan183ca3aac885 type dummy
Error: argument "mac-183ca3aac885" is wrong: "name" too long
ip link add mac-183ca3aac885 type dummy
Error: argument "mac-183ca3aac885" is wrong: "name" too long
```

Best we could prepend without further truncating the netID that still has some meaning is:

```
ip link add mac183ca3aac885 type dummy
ip link add ip183ca3aac885 type dummy
```

nerdalert replied:
We decided on using a prefix of `di-8b311a108e4f` for ipvlan (prepending di- for docker ipvlan) and `dm-1fd56f94e2f3` for macvlan (prepending dm- for docker machine) respectively.

mavenugo commented:
@nerdalert tried your patch e2e & am happy to see tagged use-case working across multi-host with appropriate configuration (in my OSX & virtual-box bridged adapter). These 2 drivers are really going to make the underlay integration extremely simple and I cant wait to get user feedback on these.

Apart from a few minor comments above, this PR gets a big :+1: from me.

mavenugo commented:
@nerdalert I have a couple of macvlan networks with and without parent interfaces. When I tried to create and use an ipvlan network, I get the following error. Which seems wrong.

```
$ sudo docker network create -d ipvlan ip300
51ceb9c3f7877e674da65861eb8a8e8770fe6a36a66d705c4f51d6a4c51b6a98

$ sudo docker run --net=ip300 -it busybox sh
WARN[9635] failed to cleanup ipc mounts:
failed to umount /var/lib/docker/containers/53dece5c8166d90457f6270fe9018b37ccd1f9c5567cfa09f45005d9734a2965/shm: no such file or directory
ERRO[9635] Handler for POST /v1.22/containers/53dece5c8166d90457f6270fe9018b37ccd1f9c5567cfa09f45005d9734a2965/start returned error: Failed creating veth2db7a06, ensure there are no macvlan networks using the same `-o parent`: operation not supported
docker: Error response from daemon: Failed creating veth2db7a06, ensure there are no macvlan networks using the same `-o parent`: operation not supported.
```

Also the same issue when I tried to create an entirely new parent interface 

```
$ sudo docker network create -d ipvlan -o parent=eth3.345 ip345
df831819e495cd0d561bc6cd1f84afdd887acdbced4c2bfdabffcbbbe02cbbcf

$ sudo docker run --net=ip345 -it busybox sh
WARN[9809] failed to cleanup ipc mounts:
failed to umount /var/lib/docker/containers/0064e630b9f4df3902c04796ba23f63d7b0b1aa60567082e092caf09512f477c/shm: no such file or directory
ERRO[9809] Handler for POST /v1.22/containers/0064e630b9f4df3902c04796ba23f63d7b0b1aa60567082e092caf09512f477c/start returned error: Failed creating vethb59a154, ensure there are no macvlan networks using the same `-o parent`: operation not supported
docker: Error response from daemon: Failed creating vethb59a154, ensure there are no macvlan networks using the same `-o parent`: operation not supported.
```

Can you pls take a look @ it ?

fredhsu commented:
I've been running some tests with an Arista ToR acting as the gateway for macvlan/ipvlan, with and without 802.1q.  So far so good!

macvlan bridge mode:
https://gist.github.com/fredhsu/63344b94fc18c87f8ed2

ipvlan L2 mode with 802.1q:
https://gist.github.com/fredhsu/38d75e7651dae8506ba2

Let me know if there are any specific tests you would like to have run against physical hardware.

nerdalert commented:
Thanks @fredhsu appreciate you testing the physical side of the equation.

nerdalert commented:
Regarding the `operation not supported.` was kernel versions < 3.19. Rather then warning in driver init(), it now fails network creates for `macvlan < 3.09` and `ipvlan < 3.19`.

mavenugo commented:
Thanks @nerdalert.

LGTM

aboch commented:
LGTM

ping @chenchun

chenchun commented:
:+1:  Introducing them in experimental.

LGTM, thanks for your work @nerdalert
